data_model/primitives/src/conststr.rs

```diff
@@ -0,0 +1,545 @@
+                PartialEq::eq(&self[..], &other[..])
+            }
+        }
+    )*};
+}
+
+impl_eq!(String, str, &str);
+
+impl Eq for ConstString {}
+
+impl Serialize for ConstString {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        serializer.serialize_str(self)
+    }
+}
+
+impl Encode for ConstString {
+    fn size_hint(&self) -> usize {
+        self.as_bytes().size_hint()
+    }
+
+    fn encode_to<W: Output + ?Sized>(&self, dest: &mut W) {
+        self.as_bytes().encode_to(dest)
+    }
+
+    fn encode(&self) -> Vec<u8> {
+        self.as_bytes().encode()
+    }
+
+    fn using_encoded<R, F: FnOnce(&[u8]) -> R>(&self, f: F) -> R {
+        self.as_bytes().using_encoded(f)
+    }
+}
+
+impl IntoSchema for ConstString {
+    fn type_name() -> String {
+        String::type_name()
+    }
+    fn schema(map: &mut MetaMap) {
+        String::schema(map);
+    }
+}
+
+/// Union representing const-string variants: inlined or boxed.
+/// Distinction between variants are achieved by setting least significant bit for inlined variant.
+/// Boxed variant should have have 4/3/2 (depending on architecture 64/32/16 bit pointer width) trailing zeros due to pointer alignment.
+#[repr(C)]
+union ConstStringData {
+    inlined: InlinedString,
+    boxed: ManuallyDrop<BoxedString>,
+}
+
+impl ConstStringData {
+    #[inline]
+    const fn is_inlined(&self) -> bool {
+        self.inlined().is_inlined()
+    }
+
+    #[allow(unsafe_code)]
+    #[inline]
+    const fn inlined(&self) -> &InlinedString {
+        // SAFETY: safe to access if `is_inlined` == `true`.
+        unsafe { &self.inlined }
+    }
+
+    #[allow(unsafe_code)]
+    #[inline]
+    const fn boxed(&self) -> &ManuallyDrop<BoxedString> {
+        // SAFETY: safe to access if `is_inlined` == `false`.
+        unsafe { &self.boxed }
+    }
+
+    #[inline]
+    fn len(&self) -> usize {
+        if self.is_inlined() {
+            self.inlined().len()
+        } else {
+            self.boxed().len()
+        }
+    }
+
+    #[inline]
+    const fn new() -> Self {
+        Self {
+            inlined: InlinedString::new(),
+        }
+    }
+}
+
+impl<T: ?Sized> AsRef<T> for ConstStringData
+where
+    InlinedString: AsRef<T>,
+    BoxedString: AsRef<T>,
+{
+    #[inline]
+    fn as_ref(&self) -> &T {
+        if self.is_inlined() {
+            self.inlined().as_ref()
+        } else {
+            self.boxed().as_ref()
+        }
+    }
+}
+
+impl<T> From<T> for ConstStringData
+where
+    T: TryInto<InlinedString>,
+    <T as TryInto<InlinedString>>::Error: Into<BoxedString>,
+{
+    #[inline]
+    fn from(value: T) -> Self {
+        match value.try_into() {
+            Ok(inlined) => Self { inlined },
+            Err(value) => Self {
+                boxed: core::mem::ManuallyDrop::new(value.into()),
+            },
+        }
+    }
+}
+
+impl Clone for ConstStringData {
+    fn clone(&self) -> Self {
```

Comment:
does overriding `clone_from` makes sense?

Reply:
@appetrosyan, you mean create `ConstString` from `&str`?

Reply:
> So I'm not sure if we need to implement this.

Rule of thumb. If we don't need it yet, and you're not sure we will, don't implement.